src/Microsoft.OData.Core/Json/JsonLightInstanceAnnotationWriter.cs

```diff
@@ -201,11 +201,13 @@ internal void WriteInstanceAnnotationsForError(ICollection<ODataInstanceAnnotati
             ODataResourceValue resourceValue = value as ODataResourceValue;
             if (resourceValue != null)
             {
+                IDuplicatePropertyNameChecker duplicatePropertyNameChecker = this.valueSerializer.GetDuplicatePropertyNameChecker();
                 this.WriteInstanceAnnotationName(propertyName, name);
                 this.valueSerializer.WriteResourceValue(resourceValue,
                     expectedType,
                     treatLikeOpenProperty,
-                    this.valueSerializer.CreateDuplicatePropertyNameChecker());
+                    duplicatePropertyNameChecker);
+                this.valueSerializer.ReturnDuplicatePropertyNameChecker(duplicatePropertyNameChecker);

```

Comment:
This really makes me think we should do the `IDisposable` thing that someone else was talking about. Don't we want to return the checker even if `WriteResourceValue` throws an exception?

Reply:
1. `IDisposable` will make the `ObjectPool` be tightly coupled with the `DuplicatePropertyNameChecker`, and this is something we are trying to avoid.

Reply:
That's a good point about returning objects even when there are exceptions.

Reply:
There's no coupling with `DuplicatePropertyNameChecker`. `ObjectPool<T>` could have a contract like:
```
public sealed class ObjectPool<T>
{
  public Lease<T> Get()
  {
    ...
  }

  private void Return(T value)
  {
    ...
  }

  public sealed class Lease<T> : IDisposable
  {
    public Lease(T value, ObjectPool<T> pool)
    {
      ...
    }

    public T Value { get; }
    public void Dispose()
    {
      this.pool.Return(this.Value);
    }
  }
}
```

Whether we actually use `IDisposable` or not, we should probably be wrapping the `ReturnDuplicatePropertyNameChecker` calls in a `finally` block (the equivalent of `using` on the proposed `Lease` type).

Reply:
Why do we need to return the object to the pool when we have an error? The writing process is already terminated. When we start writing a new response, a new ObjectPool will be created. The existing object pool will be cleaned up.

Reply:
@corranrogue9 I like the `Lease<T>` idea, but I also think it introduces some implicit coupling. Now methods that return `DuplicatePropertyNameChecker` will have to change their signature to return `Lease<DuplicatePropertyNameChecker>`:
```c#
using Lease<DuplicatePropertyNameChecker> leasedChecker = ...GetDuplicateNameChecker();
DuplicateNameChecker duplicateNameChecker = leastedChecker.Value;
```

 Maybe this indirection is better than calling `Return()` manually. But there would probably still be cases where you can't use `using` and have to call `Dispose()` manually.

But I think Ken also makes a good point. We're not holding an external resource. The risk of not returning an object is relatively low, it will still be eventually garbage collected. And as Ken mentions, the lifetime of the pool is the request, so if a request fails, the pool and objects not yet returned will be eventually cleaned up.

What if we code that tries to verify correctness of the object pool at debug or test time. Maybe debug assertions that keep track of how many objects were created and returned and verified before the request is disposed. And test cases could also verify this. I think this avoids the need for `Lease<T>` or throwing exceptions when the nesting level of the request exceeds the size of the object pool.

Reply:
The other concern I have with `Lease<T>` approach is that it will add an extra allocation each an object is requested from the pool. So you'll end up having as many `Lease<T>` as nested objects in the response, which is the exact thing this PR is trying to avoid.

Reply:
I'm not too attached to the `Lease` idea, a note on the allocations is that a `Lease` *could* be a `struct` I think? I'm going to resolve though